src/CloudNative.CloudEvents.SystemTextJson/JsonEventFormatter.cs

```diff
@@ -108,41 +108,32 @@ public JsonEventFormatter(JsonSerializerOptions serializerOptions, JsonDocumentO
             DocumentOptions = documentOptions;
         }
 
-        public override async Task<CloudEvent> DecodeStructuredModeMessageAsync(Stream data, ContentType contentType, IEnumerable<CloudEventAttribute> extensionAttributes)
-        {
-            data = data ?? throw new ArgumentNullException(nameof(data));
-            var encoding = contentType.GetEncoding();
+        public override Task<CloudEvent> DecodeStructuredModeMessageAsync(Stream data, ContentType contentType, IEnumerable<CloudEventAttribute> extensionAttributes) =>
+            DecodeStructuredModeMessageImpl(data, contentType, extensionAttributes, true);
```

Comment:
Should we await this call and mark the method as async?

Reply:
No need - the task returned by the "impl" method is fine already. (If that method had been implemented without using `async`, there'd be some potential for difference in terms of timing for exceptions etc, but as it is, they'll behave the same way.)

Reply:
I believe the best practice is to always await the task in a public API to provide an understandable call stack.

Reply:
Hmm... happy to do that if it's really best practice (not one I've come across before for this situation), but I don't *think* the stack trace will be hard to understand here - it feels like it would just be adding a perf hit.

Do you have any links to docs about this? Always eager to learn more about best practices. If there's something that's MSFT internal, feel free to say so and I'll take it on trust :)

Reply:
https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#prefer-asyncawait-over-directly-returning-task

Reply:
Cool, thanks. Will implement tomorrow - and look at ways of simplifying the data checks as well. (The fact that it wasn't immediately clear to you is enough signal that it could do with work.)